### PR TITLE
backend: fix prepare_read behavior

### PR DESCRIPTION
**Pull request by elinorbgr**
This should fix the multithreading issues that have been encountered.

cc @kchibisov

### CODE REVIEW

File: wayland-client/src/conn.rs [hunk 1]
```diff
@@ -136,7 +136,7 @@ impl Connection {
     ///
     /// If you don't need to manage multiple event sources, see
     /// [`blocking_dispatch()`](EventQueue::blocking_dispatch) for a simpler mechanism.
-    pub fn prepare_read(&self) -> Result<ReadEventsGuard, WaylandError> {
+    pub fn prepare_read(&self) -> Option<ReadEventsGuard> {
```
Comment (kchibisov) [inline wayland-client/src/conn.rs:139]: Should it be marked as `#[must_use]`?

File: wayland-client/src/conn.rs [hunk 2]
```diff
@@ -162,16 +162,12 @@ impl Connection {
...
-            // first, prepare the read
-            let guard = self.backend.prepare_read()?;
-
-            // If another thread processed events prior to prepare_read, we might already be done.
-            if done.done.load(Ordering::Relaxed) {
-                break;
+            if let Some(guard) = self.backend.prepare_read() {
+                dispatched += blocking_read(guard)?;
+            } else {
+                dispatched += self.backend.dispatch_inner_queue()?;
```
Comment (kchibisov) [inline wayland-client/src/conn.rs:168]: Hm, but you could have a blocking dispatch of the queue here as well? Non-blocking roundtrip doesn't sound like it'll work reliably or I don't understand something?

**Review by kchibisov — CHANGES_REQUESTED** [review]
Probably the same about `#[must_use]` for other methods. Also, maybe we should use `Result` instead, but use some `io` type to indicate that lock can't be acquired? The mutex stuff uses `Result` and it's more intuitive than the `Option`.

File: wayland-client/src/conn.rs (hunk 1 above)
Comment (elinorbgr) [reply wayland-client/src/conn.rs:139]: good point :+1:

File: wayland-protocols/CHANGELOG.md [hunk 3]
```diff
@@ -5,7 +5,7 @@
 ### Breaking changes
 
 - Bump bitflags to 2.0
-- Updated wayland-backend to 0.2
+- Updated wayland-backend to 0.3
```
Comment (kchibisov) [inline wayland-protocols/CHANGELOG.md:8]: This looks arbitrary and unrelated to this PR, should be submitted separately.

File: wayland-client/src/conn.rs (hunk 2 above)
Comment (elinorbgr) [reply wayland-client/src/conn.rs:168]: This is in a loop, so it'll just try again at the next iteration.

File: wayland-protocols/CHANGELOG.md (hunk 3 above)
Comment (elinorbgr) [reply wayland-protocols/CHANGELOG.md:8]: Well, this PR introduces breaking changes that cascade to all other crates, so it seems appropriate to add the associated changelog entries in the same PR.

File: wayland-client/src/conn.rs (hunk 2 above)
Comment (kchibisov) [reply wayland-client/src/conn.rs:169]: Ah, yeah, sure, I just know that there's mo efficient API to do the exact same thing, but I think it's not an option for us. The `wl_display_dispatch_queue` https://wayland.freedesktop.org/docs/html/apb.html#Client-classwl__display_1ae027b09801474ac7c6b0f1ef25ff6e17

File: wayland-protocols/CHANGELOG.md (hunk 3 above)
Comment (kchibisov) [reply wayland-protocols/CHANGELOG.md:8]: But you don't actually bump the version here, so it should be done in a separate commit.

File: wayland-client/src/conn.rs (hunk 2 above)
Comment (elinorbgr) [reply wayland-client/src/conn.rs:169]: That function in wayland-client does mostly exactly the same thing as `wl_display_roundtrip`, but implemented in terms of wayland-backend's API, so I don't really get your point?

File: wayland-protocols/CHANGELOG.md (hunk 3 above)
Comment (elinorbgr) [reply wayland-protocols/CHANGELOG.md:8]: ok

**Comment by elinorbgr** [comment]
> Also, maybe we should use Result instead, but use some io type to indicate that lock can't be acquired?

Not sure what we would gain from this? It feels to me that this `Option<_>` reflects best the behavior of this method, changing it to a `Result` we would just have an error type identical to `()`, what would be the point then?

File: wayland-client/src/conn.rs (hunk 2 above)
Comment (kchibisov) [reply wayland-client/src/conn.rs:169]: `wl_display_roundtrip` uses `poll` under the hood to not do the loop.

**Review by kchibisov — APPROVED** [review]

File: wayland-client/src/conn.rs (hunk 2 above)
Comment (elinorbgr) [reply wayland-client/src/conn.rs:169]: What... ?

`wl_display_roundtrip` does have a loop: https://gitlab.freedesktop.org/wayland/wayland/-/blob/main/src/wayland-client.c#L1402

and `blocking_read` uses `poll` internally: https://github.com/Smithay/wayland-rs/blob/master/wayland-client/src/conn.rs#L224-L245

Comment (kchibisov) [reply wayland-client/src/conn.rs:169]: @elinorbgr if you look closely what their code is and what we have, they use blocking dispatch. However we use it only in `read_guard` branch, and the other branch is using regular non-blocking dispatch.

So while they have a `loop`, the `wait`, but we `busy loop` from the `else`. The blocking read is fine, the `else` is a busy loop. `dispatch_inner_queue` is a non_blocking API from what I can, see, but libwayland always using blocking API, with `dispatch_queue`.

Comment (kchibisov) [reply wayland-client/src/conn.rs:169]: After reading the libwayland impl it's pretty much the same, when their docs are not really clear what exactly blocks and how it works.

**Pull request merged**